kid0012.md

```diff
@@ -0,0 +1,62 @@
+8. If there is duplicity, super watchers detect and report it within 2 minutes of its manifestation in diverged witness state.
+
+## Requirements
+
+9. Witnesses MUST implement an endpoint that describes which of the following features they support, so that their behavior is discoverable. (Or should the TEL of the witness simply contain a commitment to support certain features?)
+
+10. Witnesses SHOULD support an efficient, comprehensive polling feature required by super watchers. Essentially, super watchers will ask, "What changes have you recorded since your &lt;checkpoint X&gt;?", and witnesses implementing the feature MUST respond with a batch of changes.
+
+    ![diagram of polled witnesses](assets/polled-witnesses.png)
+
+    This is probably implemented by having witnesses maintain a data structure called a compact optimized sparse merkle tree, which makes it easy to reference versions of data by the root hash of the merkle tree. Super watchers who want to poll can say to a witness, "What has changed since X was the hash of your merkle tree?" and receive a very terse response that allows them to bring their own copy of the same data up-to-date.
+
+11. Witnesses SHOULD throttle polling requests.
+
+12. Witnesses MAY also restrict who can poll. Exactly how they do this is out of scope. One obvious way to restrict is by requiring polling requests to be signed by an AID on an allow list.
+
+13. Witnesses MAY support a push notification feature as well. If they do, they allow themselves to be configured so that they push notifications of changes in the KEL/TEL of any of their serviced AIDs to an arbirary endpoint (a super watcher).
+
+![pushing witnesses](assets/pushing-witnesses.png)
+
+The super watcher benefits because it learns of changes with lower latency and greater efficiency than it would if it polled. However, the witness accepts a burden to provide this feature. For this reason, witnesses will probably only allow the feature if the watcher in question belongs to the same or an allied owner.
+
+14. If witnesses support this push feature, they SHOULD push to only one location, delivery SHOULD be best-effort, and witnesses SHOULD contain no retry logic or state tracking. This preserves the simplicity of witnesses, and their focus on recording data, not moving it.
+
+15. If witnesses support this push feature, they SHOULD also support aggregate push. This is a network optimization, where updates that occur within a configurable timeframe (e.g., 10 seconds) are all pushed as a single notification.
+
+16. Duplicity detection. A super watcher SHOULD be a juror. A super watcher MAY be a judge. We need to work on some general recipes that a judge could follow.
+
+17. A witness MUST expose an endpoint that announces any watchers that poll it, so the watcher's reputations can be computed.
+
+18. A witness MUST expose an endpoint that announces if it is pushing to a watcher, so the watcher's reputation can be computed.
```

Comment:
I guess this is the same question I had above.  It seems like publicly available watchers (or super watchers) would want this but it does seem like there might be a use case for discrete watchers who may not wish to be enumerated publicly.

Reply:
Also a good improvement that I'll try to incorporate in an update.